prov/verbs/src/verbs_mr.c

```diff
@@ -57,6 +57,64 @@ static struct fi_ops vrb_mr_fi_ops = {
 	.ops_open = fi_no_ops_open,
 };
 
+#if HAVE_SYNAPSEAI
+static struct ibv_mr *vrb_reg_synapseai_dmabuf(struct ibv_pd *pd,
+					       const void *buf, size_t len,
+					       int vrb_access)
+{
+	int dmabuf_fd;
+	int ret;
+	struct ibv_mr *mr;
+	int saved_errno = 0;
+	enum { TRY, ALWAYS, NEVER };
+	static int failover_policy = TRY;
+
+	if (failover_policy == ALWAYS)
+		goto failover;
+
+	ret = synapseai_get_dmabuf_fd((uint64_t) buf, (uint64_t) len,
+				      &dmabuf_fd);
+	FI_DBG(&vrb_prov, FI_LOG_MR,
+	       "synapseai returned dmabuf FD %d for addr=[%p], size=[%zd] \n",
+	       dmabuf_fd, buf, len);
+	if (ret != FI_SUCCESS) {
+		FI_WARN(&vrb_prov, FI_LOG_MR,
+			"Unable to get dmabuf fd for Gaudi device buffer \n");
+		return NULL;
+	}
+
+	mr = ibv_reg_dmabuf_mr(pd, 0, len, (uint64_t) buf /* iova */,
+			       (int) (uintptr_t) dmabuf_fd /* dmabuf fd */,
+			       vrb_access);
+	if (!mr && failover_policy == TRY && vrb_gl_data.peer_mem_support) {
+		saved_errno = errno;
+		goto failover;
+	}
+
+	failover_policy = NEVER;
+	return mr;
+
+failover:
+	mr = ibv_reg_mr(pd, buf, len, vrb_access);
+	if (!mr) {
+		if (saved_errno) {
+			FI_INFO(&vrb_prov, FI_LOG_MR,
+				"Failover failed: ibv_reg_mr(%p, %zd) error "
+				"%d\n",
+				buf, len, errno);
+			errno = saved_errno;
+		}
+		return NULL;
+	}
+	if (failover_policy == TRY) {
+		failover_policy = ALWAYS;
+		FI_INFO(&vrb_prov, FI_LOG_MR,
+			"Failover on: ibv_reg_dmabuf_mr() ==> ibv_reg_mr()\n");
+	}
+	return mr;
+}
+#endif
```

Comment:
This function is mostly the same as vrb_reg_ze_dmabuf().  The duplication can be eliminated by:
(1) add an `iface` argument to the call;
(2) add `get_dmabuf_fd` method to `ofi_hmem_ops` and define `ofi_hmem_get_dmabuf_fd()`;
(2) define `get_base_addr` method for synapseai;
(3) use `ofi_hmem_get_dmabuf_fd()` and `ofi_hmem_get_base_addr()` instead of iface specific calls;
(4) change the function name to something like `vrb_reg_dmabuf`.

Reply:
See #9466 for step (2) mentioned in previous comment.

Reply:
@nbeer-habana I noticed that in your patch, the `offset` passed to `ibv_reg_dmabuf_mr` is always 0. Does Synapseai mandate that the `buf` here must be the starting address of the dmabuf object? If so, is it checked inside the `hcclLookupDMABuff` call? If not, is there a way to get the base address of the dmabuf object so that the offset can be calculated?

Reply:
@j-xiong Yes, SynapseAI assumes that `buf` is the starting address of the dmabuf object. We don't see a use-case where we would need to register part of a dmabuf object.